runtime/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -3018,56 +3143,112 @@ else if (mqttPublishHeaderRO.retained && !retainAvailable(capabilities))
                 onDecodeError(traceId, authorization, reasonCode);
                 decoder = decodeIgnoreAll;
             }
-            else
+
+            if (unreleasedPacketIds.contains(packetId))
             {
-                if (!unreleasedPacketIds.contains(mqttPublishHeaderRO.packetId))
+                switch (version)
                 {
-                    if (mqttPublishHeaderRO.qos == 2)
-                    {
-                        unreleasedPacketIds.add(mqttPublishHeaderRO.packetId);
-                    }
+                case 4:
+                    doEncodePubrecV4(traceId, authorization, packetId);
+                    break;
+                case 5:
+                    doEncodePubrecV5(traceId, authorization, packetId);
+                    break;
+                }
+            }
+
+            doSignalKeepAliveTimeout(traceId);
+        }
 
+        private void onDecodePublishPayload(
+            long traceId,
+            long authorization,
+            int reserved,
+            int packetId,
+            OctetsFW payload,
+            int offset,
+            int limit)
+        {
+            int reasonCode = SUCCESS;
+
+            if (mqttPublishHeaderRO.payloadFormat.equals(MqttPayloadFormat.TEXT) && invalidUtf8(payload))
```

Comment:
What if `mqttPublishHeaderRO` is reused by a different client, can it cause a bug?
Prefer not to refer to factory-scoped state from client-specific methods other than on stack use.